### ISSUE TITLE
Doesn't protect against Azure Facial Recognition in my test

### ISSUE THREAD

**Issue by minha12** [Shawn-Shan/fawkes#59]
I used 10 photos in IMM dataset (https://www2.imm.dtu.dk/~aam/datasets/datasets.html). I tested the software by following steps:

- Creating the cloak images according to your codes with option '-mode' set to 'high'. It took 2 hours for generating 10 images in my local PC!!!. The result's in the link: https://github.com/minha12/test_fawkes/tree/master/imm/train_high
- Using Azure Facial Recognition demo test in the link: https://azure.microsoft.com/en-us/services/cognitive-services/face/#demo
- Upload cloak image (first) and then uncloak image from the same person (different from the original image used for cloaking).

Result: For 10 images I tried, the 'confidence' is still high, approximately 0.9 for all the case. 

I also did the same test with face_recognition module (which is based on Dlib), the result is still the same. I don't what should I do to achieve the same level of protection as described in the paper?

**Comment by emjoywill** [comment]
A few things:

0. Fawkes is designed to fool _facial recognition_ models, not _face detection_ models. The tutorial you linked seems to be doing the latter, not the former. However, I may be misinterpreting. 
1. If your experiment is actually on face recognition, did you use exclusively cloaked images of you to train a new model before evaluating on uncloaked photos of you? Our experiments use Azure's "train your own model" service, so you need to be sure this is what you're doing. 
2. If you did train on cloaked images, did you have other people in the training dataset? Cloaks work best when they are part of a much-larger facial recognition dataset. The presence of other classes in the dataset ensure that feature space is sufficiently filled by other users. Without other classes in the model, the model associates basically all of feature space with you, so even though uncloaked images show up in a different part of feature space, they are still associated with the cloaked class.

**Comment by minha12** [comment]
Thanks for your explanation and tips.  I have some experience with Microsoft Azure Face Recognition System (FR for short)  so I was surprised to hear your research result. Nice idea, indeed, but after running some tests, I couldn't convince myself about the 100% protection rate as in your paper. So, I wrote some scripts to train a FR model with Azure FR API in this link:
https://github.com/minha12/test_fawkes/blob/ca5de7957e26d0eecafd2eebc63c640c35170453/Azure_FR/Azure_FR.ipynb
This is by no mean a comprehensive test, just a test with a small IMM dataset (40 identities). Currently, the result show that with the -mode = min. Majority of the identities still be re-identified with confidence around 0.7 - 0.8 (see the end of the notebook above). 

0. The demo site has a section called 'Face verification' if you scroll down. In here, user can upload two images to compare at a time. This is just a quick test.

1. Yes I used the same service.

2. I did train with only cloak images as shown in the link above but the result doesn't look as pretty (as yours).

Depending on your definition of 'training'. Nowadays, most of FR systems use triplet-loss to calculate embedding. If you mean 'training' is training from scratch the embedding generator (e.g. FaceNet) then probably this method can some how reduce the FR accuracy. However, what prevents the one who's building a FR to use the pre-train embedding generator? What I mean is that training a FR nowadays is simply calculating a database of embbeding from training images. My test using FaceNet to calculating embedding on clock images didn't have a good protection rate as well.

**Comment by Shawn-Shan** [comment]
Hi, 
Thanks for the note and I will take a deeper look at Azure demo. We used the Azure API back in Jan 2020, which has a train endpoint that supposedly train a model for you. We know that they probably trained some sort of a model because the result is stochastic every time. As for the FaceNet comment, we didn't really test against FaceNet or autoencoder in the paper. It is certainly interesting to look at those as more and more system start adopting them. But I believe the intuition is the same, and we might need to use a FaceNet model for optimization rather than a traditional classification model. 

I look take a look at Azure's new API and their demo, as well as some FaceNet models. Thanks for surfacing this!!
